Register application in Azure AD

maartenba opened the pull request:
Fixes #343

This PR provides a way to register the current application as an Azure Active Directory application.
* There is a **Tools \| Azure \| Register application with Azure AD** menu entry to do this
* When using the ASP.NET Core application template and selecting authentication with Azure Active Directory, a notification will be shown inviting the user to finish setting things up

Example flow:

1) User creates new ASP.NET Core application and uses an _Organizational account_ (Azure Active Directory):
![image](https://user-images.githubusercontent.com/485230/90043493-7a35b280-dccc-11ea-989a-d767ac0b028f.png)

2) Rider detects the defaults from the project template, and provides a notification in the editor:
![image](https://user-images.githubusercontent.com/485230/90043547-8c175580-dccc-11ea-9e77-631eaa92ffbc.png)

3) When invoked, application details can be updated:
![image](https://user-images.githubusercontent.com/485230/90043590-99344480-dccc-11ea-9754-b6e46017340a.png)

4) After clicking OK, Rider rgisters the application in Azure Active Directory, and updates `appsettings.json` with the required details:
![image](https://user-images.githubusercontent.com/485230/90043683-b9640380-dccc-11ea-80ba-29ceee98e00f.png)

maartenba committed:
Register application in Azure AD #343

maartenba committed:
Move managedidentity package to identity package

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/identity/ad/notifications/DefaultAzureAdApplicationRegistrationNotificationProvider.kt:
```diff
@@ -0,0 +1,62 @@
+import org.jetbrains.plugins.azure.identity.ad.actions.RegisterApplicationInAzureAdAction
+import org.jetbrains.plugins.azure.identity.ad.appsettings.AppSettingsAzureAdSectionManager
+
+class DefaultAzureAdApplicationRegistrationNotificationProvider : EditorNotifications.Provider<EditorNotificationPanel>() {
```
The notification does not have "Ignore" option for cases when user do not care about registering an app in Azure AD.

maartenba replied:
Fixed.

sdubov commented:
Getting an exception when trying to Register Azure AD application:
```
com.microsoft.azure.management.graphrbac.GraphErrorException: Status code 403, {"odata.error":{"code":"Authorization_RequestDenied","message":{"lang":"en","value":"Insufficient privileges to complete the operation."},"requestId":"6e61588e-d466-4a9c-a9e4-c08d932165eb","date":"2020-08-12T17:34:23"}}
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at com.microsoft.rest.ServiceResponseBuilder.build(ServiceResponseBuilder.java:122)
	at com.microsoft.azure.AzureResponseBuilder.build(AzureResponseBuilder.java:56)
	at com.microsoft.azure.management.graphrbac.implementation.ApplicationsInner.createDelegate(ApplicationsInner.java:194)
	at com.microsoft.azure.management.graphrbac.implementation.ApplicationsInner.access$000(ApplicationsInner.java:45)
	at com.microsoft.azure.management.graphrbac.implementation.ApplicationsInner$2.call(ApplicationsInner.java:181)
	at com.microsoft.azure.management.graphrbac.implementation.ApplicationsInner$2.call(ApplicationsInner.java:177)
	at rx.internal.operators.OnSubscribeMap$MapSubscriber.onNext(OnSubscribeMap.java:69)
	at retrofit2.adapter.rxjava.CallArbiter.deliverResponse(CallArbiter.java:120)
	at retrofit2.adapter.rxjava.CallArbiter.emitResponse(CallArbiter.java:102)
	at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(CallExecuteOnSubscribe.java:46)
	at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(CallExecuteOnSubscribe.java:24)
	at rx.Observable.unsafeSubscribe(Observable.java:10327)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:48)
	at rx.internal.operators.OnSubscribeMap.call(OnSubscribeMap.java:33)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:48)
	at rx.internal.operators.OnSubscribeLift.call(OnSubscribeLift.java:30)
	at rx.Observable.subscribe(Observable.java:10423)
	at rx.Observable.subscribe(Observable.java:10390)
	at rx.observables.BlockingObservable.blockForSingle(BlockingObservable.java:443)
	at rx.observables.BlockingObservable.single(BlockingObservable.java:340)
	at com.microsoft.azure.management.graphrbac.implementation.ApplicationsInner.create(ApplicationsInner.java:127)
	at org.jetbrains.plugins.azure.identity.ad.actions.RegisterApplicationInAzureAdAction$fetchDataAndShowDialog$1.run(RegisterApplicationInAzureAdAction.kt:229)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:935)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsync$5(CoreProgressManager.java:442)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:235)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:170)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:629)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:581)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:157)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:235)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:668)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:665)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:665)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

maartenba committed:
Register application in Azure AD #343 - Add learn more link to dialog

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/identity/ad/ui/RegisterApplicationInAzureAdDialog.kt:
```diff
@@ -0,0 +1,176 @@
+
+        hideableRow(RiderAzureBundle.message("dialog.identity.ad.register_app.more_options")) {
+            row {
+                val label = JLabel(RiderAzureBundle.message("dialog.identity.ad.register_app.client_id"))
```
Do we need to add validation for Client ID. I'm getting `Status code 400, "Unsupported or invalid query filter clause specified for property 'appId' of resource 'Application'."` error when Client ID is not a valid Guid.

maartenba replied:
Fixed.

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/identity/ad/ui/RegisterApplicationInAzureAdDialog.kt:
```diff
@@ -0,0 +1,176 @@
+import org.jetbrains.plugins.azure.isValidUrl
+import javax.swing.JLabel
+
+class RegisterApplicationInAzureAdDialog(project: Project, val model: RegistrationModel)
```
To discuss - maybe change [Ok] button name to [Register]?

maartenba replied:
Fixed.

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/identity/ad/ui/RegisterApplicationInAzureAdDialog.kt:
```diff
@@ -0,0 +1,176 @@
+            label()
+
+            cell {
+                textField({ model.originalCallbackUrl },
```
The default Callback URL is loaded when I change `appsettings.json` and reopen the dialog to re-register an app.

maartenba replied:
Fixed.

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/identity/ad/ui/RegisterApplicationInAzureAdDialog.kt:
```diff
@@ -0,0 +1,176 @@
+                            model.updatedDomain = it
+                            model.hasChanges = true
+                        })
+                        .withValidationOnInput { validateNotEmpty(it) }
```
Skip trailing spaces?

maartenba replied:
Fixed.

maartenba committed:
Register application in Azure AD #343 - Fix PR comments

maartenba committed:
Register application in Azure AD #343 - Show notification when administrator has disabled app access

maartenba commented:
* Added support for `appsettings.{environment}.json`
* Added notification instead of exception for [this case](https://github.com/JetBrains/azure-tools-for-intellij/pull/349#issuecomment-673036981)
* Fixed review comments
* Cosmetics

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/RiderAzureUtils.kt:
```diff
@@ -22,8 +22,34 @@
+    }
+}
+
+fun String?.isValidUUID(): Boolean {
```
Could you please cover this with some tests

maartenba replied:
Fixed.

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/identity/ad/appsettings/AppSettingsAzureAdSection.kt:
```diff
@@ -0,0 +1,37 @@
+) {
+
+    // Is this instance the default project template content (generated by dotnet new)?
+    fun isDefaultProjectTemplateContent() =
```
Could you please cover this with some tests?

maartenba replied:
Fixed.

sdubov reviewed (approved):
Looks good. It would be nice to cover helper functions (please see comments) with tests.
Also, could you please update platformPlugin.xml `<change-notes>` section to not make a separate commit?

maartenba committed:
Register application in Azure AD #343 - Add some tests

maartenba committed:
Register application in Azure AD #343 - Add some tests
